cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala

```diff
@@ -119,166 +124,487 @@ class RandomTraverser(graph: Graph, dir: GraphDir, homeNodeIds: Seq[Int],
-  }
+  // can't limit number of nodes added to the queue, because we skip some nodes
+  override def limitAddedToQueue(nodes: Seq[Int]): Seq[Int] = nodes
+
+  override def shouldBeEnqueued(color: Walk.NodeColor.Color): Boolean = color != Walk.NodeColor.Visited
 }
 
-class DepthFirstTraverser(graph: Graph, dir: GraphDir, homeNodeIds: Seq[Int],
-    onlyOnce: Boolean)
-    extends DistanceTraverser[Int](graph, homeNodeIds, onlyOnce) {
 
-  // stack keeps a tuple (node, child# of higher node in stack). Hence, top node
-  // of stack always has child# = -1
-  case class NodeDesc(val id: Int, val childOff: Int)
+/**
+ * A trait to be mixed in to the DepthFirstTraverser that keeps track of visiting distance
+ * from `homeNodeIds`.
+ *
+ * Note that DepthTracker for the DFS case returns the depth of first seeing a particular
+ * node. It does not have to be the same as the depth at which the node is visited.
+ */
+trait PathLengthTracker extends DepthFirstTraverser {
+  private lazy val nextVisitDistanceTracker = new IntInfoKeeper(false)
+  private lazy val visitDistanceTracker = new IntInfoKeeper(true)
+
+  abstract override protected def enqueue(nodes: Seq[Int], from: Option[Int]): Unit = {
+    val fromDistance = from.flatMap(visitDistanceTracker.infoOfNode).getOrElse(-1)
+    nodes foreach { id =>
+      nextVisitDistanceTracker.recordInfo(id, fromDistance + 1)
+    }
+    super.enqueue(nodes, from)
+  }
 
-  private val stack = new ObjectArrayList[NodeDesc]
+  override def visitNode(node: Node) {
+    visitDistanceTracker.recordInfo(node.id, nextVisitDistanceTracker.infoOfNode(node).get)
+    super.visitNode(node)
+  }
 
-  init(-1)
+  def distance(id: Int) = visitDistanceTracker.infoOfNode(id)
+}
+
+/**
+ * Trait to be mixed in to the BFS/DFS traverser to add discovery time of a node.
+ */
+trait DiscoveryTimeTracker extends QueueBasedTraverser {
+  private[this] var time: Int = _ // automatically initialized to 0 before first enqueue
 
-  protected def storeForVisit(id: Int, info: Int) { stack.add(NodeDesc(id, info)) }
+  private lazy val discoveryTime = new IntInfoKeeper(true)
 
-  private def firstNotYetVisitedNode(nd: Node, off: Int): Option[NodeDesc] = {
-    assert(off != -1)
-    if (off == nd.neighborCount(dir)) {
-      // no more children of this node left, go to its parent
-      if (stack.isEmpty) None else {
-        val topel = stack.remove(stack.size - 1)
-        firstNotYetVisitedNode(getExistingNodeById(graph, topel.id), topel.childOff + 1)
-      }
-    } else {
-      val childId = nd.neighborIds(dir)(off)
-      if (onlyOnce && seenBefore(childId)) {
-        // try the next child
-        firstNotYetVisitedNode(nd, off + 1)
-      } else {
-        storeForVisit(nd.id, off)
-        Some(NodeDesc(childId, distance(nd.id).get + 1))
-      }
+  abstract override protected def enqueue(nodes: Seq[Int], from: Option[Int]): Unit = {
+    nodes.foreach {
+      node =>
+        if (discoveryTime(node).isEmpty) {
```

Comment:
do need to check whether to increase time. I thought it's more clear to check it before recording.